dm/pkg/binlog/pos_finder.go

```diff
@@ -0,0 +1,369 @@
+}
+
+func (r *binlogPosFinder) checkTransactionBeginEvent(ev *replication.BinlogEvent) (bool, error) {
+	// we find the timestamp at transaction boundary
+	// When there are GTID events in this binlog file, we use GTID event as the start event, else:
+	// for DML
+	// 		take a 'BEGIN' query event as the start event
+	//		XID event or a 'COMMIT' query event as the end event
+	// for DDL
+	// 		one single query event acts as both the start and end event
+	var transactionBeginEvent bool
+	switch ev.Header.EventType {
+	case replication.FORMAT_DESCRIPTION_EVENT:
+		_, err := r.parser.Parse(ev.RawData)
+		if err != nil {
+			return false, err
+		}
+	case replication.GTID_EVENT, replication.ANONYMOUS_GTID_EVENT, replication.MARIADB_GTID_EVENT:
+		// since 5.7, when GTID not enabled, mysql add a anonymous gtid event. we use this to avoid parsing query event
+		r.everMetGTIDEvent = true
+		transactionBeginEvent = true
+	case replication.QUERY_EVENT:
+		if !r.everMetGTIDEvent {
+			// user may change session level binlog-format=statement, but it's an unusual operation, so we parse it every time
+			// In MySQL 5.6.x without GTID, the timestamp of BEGIN is the timestamp of the first statement in the transaction,
+			// not the commit timestamp of the transaction.
+			// To simplify implementation, we use timestamp of BEGIN as the transaction timestamp,
+			// but this may cause some transaction with timestamp >= target timestamp be skipped.
+			// TODO maybe add backtrace to support this case later
+			ev2, err := r.parser.Parse(ev.RawData)
+			if err != nil {
+				return false, err
+			}
+			e := ev2.Event.(*replication.QueryEvent)
+			switch string(e.Query) {
+			case "BEGIN":
+				transactionBeginEvent = true
+				r.inTransaction = true
+			case "COMMIT": // MyISAM use COMMIT to end transaction
+				r.inTransaction = false
+			default:
+				if !r.inTransaction {
+					// DDL
+					transactionBeginEvent = true
+				}
+			}
+		}
+	case replication.XID_EVENT:
+		r.inTransaction = false
+	}
+	return transactionBeginEvent, nil
+}
+
+// FindByTs get binlog location of first event or transaction with timestamp >= ts
+// go-mysql has BinlogStreamer.GetEventWithStartTime, but it doesn't fit our need. And we need to support relay log.
+// if posType != AboveUpperBoundBinlogPos, then location is the target location we want.
+// if posType == BelowLowerBoundBinlogPos, master binlog may have purged.
+func (r *binlogPosFinder) FindByTs(ts int64) (*Location, PosType, error) {
+	r.tctx.L().Info("target timestamp", zap.Int64("ts", ts))
+
+	if err := r.initTargetBinlogFile(ts); err != nil {
+		return nil, InvalidBinlogPos, err
+	}
+
+	targetTs := uint32(ts)
+	position := mysql.Position{Name: r.targetBinlog.name, Pos: FileHeaderLen}
+	gtidSet := gtid.MinGTIDSet(r.flavor)
+
+	binlogReader, err := r.startSync(position)
+	if err != nil {
+		return nil, InvalidBinlogPos, err
+	}
+	defer binlogReader.Close()
+	for {
+		ev, err := binlogReader.GetEvent(r.tctx.Ctx)
```

Comment:
how about add a new ctx para to this func? this can help caller to cancel this operation if they want to, as this func may block for a long time

Reply:
caller can just cancel this `r.tctx.Ctx` which is passed from caller.